### PR TITLE
fix database dynamic labels

### LINKED ISSUE #7198: `dynamic_labels` are not found in DB RBAC
### Description

It seems that `static_labels` are taken into account for RBAC, but `dynamic_labels` are ignored.

**What happened**: 

I have a role which defines:
```
kind: role
metadata:
  id: 1622862508385742000
  name: test-db-access
spec:
  allow:
    db_labels:
      test: somevalue
```

If I define a node via:
```
db_service:
  enabled: yes
  databases:
    - name: "postgres-elegantly"
      protocol: "postgres"
      uri: "postgres-elegantly.example.net:5432"
      ca_cert_file: /etc/ssl/certs/teleport-ca.crt
      static_labels:
      	test: somevalue
      dynamic_labels:
        - name: "hostname"
          command: ["hostname"]
          period: 1h
```
...then I can connect to the db:
```
$ psql "service=teleport.example.net-postgres-elegantly" -c 'select 1'
 ?column?
----------
        1
(1 row)
```

If I define it as:
```
db_service:
  enabled: yes
  databases:
    - name: "postgres-elegantly"
      protocol: "postgres"
      uri: "postgres-elegantly.example.net:5432"
      ca_cert_file: /etc/ssl/certs/teleport-ca.crt
      dynamic_labels:
        - name: test
          command: ["echo", "-e", "somevalue"]
          period: 1m
```
...then I consistently get:
```
$ psql "service=teleport.example.net-postgres-elegantly" -c 'select 1'
psql: error: access to database denied
```
And the db service reports:
```
ERRO [DB:SERVIC] Failed to handle connection. addr:172.31.92.97:3024 error:[
ERROR REPORT:
Original Error: *trace.AccessDeniedError access to database denied
Stack Trace:
	/go/src/github.com/gravitational/teleport/lib/services/role.go:1934 github.com/gravitational/teleport/lib/services.RoleSet.CheckAccessToDatabase
	/go/src/github.com/gravitational/teleport/lib/srv/db/postgres/engine.go:182 github.com/gravitational/teleport/lib/srv/db/postgres.(*Engine).checkAccess
	/go/src/github.com/gravitational/teleport/lib/srv/db/postgres/engine.go:97 github.com/gravitational/teleport/lib/srv/db/postgres.(*Engine).HandleConnection
	/go/src/github.com/gravitational/teleport/lib/srv/db/server.go:401 github.com/gravitational/teleport/lib/srv/db.(*Server).handleConnection
	/go/src/github.com/gravitational/teleport/lib/srv/db/server.go:367 github.com/gravitational/teleport/lib/srv/db.(*Server).HandleConnection
	/go/src/github.com/gravitational/teleport/lib/reversetunnel/transport.go:234 github.com/gravitational/teleport/lib/reversetunnel.(*transport).start
	/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit
User Message: access to database denied] db/server.go:369
```

In both cases, `tsh db ls` shows:
```
Name                                          Description Labels                                                                Connect
--------------------------------------------- ----------- --------------------------------------------------------------------- --------------------------------------------------------
> postgres-elegantly (user: example, db: example)             hostname=postgres-elegantly.example.net,test=somevalue psql "service=teleport.example.net-postgres-elegantly"
```

**What you expected to happen**:
The dynamic label should also have allowed access.

### Reproduction Steps
As above.

### Server Details
- Teleport version (run `teleport version`): `Teleport v6.2.1 git:v6.2.1-0-gbada2617a go1.15.5`
- Server OS (e.g. from `/etc/os-release`): Ubuntu 20.04.2
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): AWS
- Additional details:

### Client Details
- Tsh version (`tsh version`): `Teleport v6.0.3 git: go1.16.3`
- Computer OS (e.g. Linux, macOS, Windows): macOS
- Browser version (for UI-related issues):
- Installed via (e.g. apt, yum, brew, website download): brew
- Additional details:

### Debug Logs
Please include or attach debug logs, when appropriate. Obfuscate sensitive information!
- Start Teleport with --debug flag (`teleport --debug`)

From the database server:
```
[DB:SERVIC] DEBU Transport request: teleport-transport. leaseID:1 target:teleport.example.net:3024 reversetunnel/agent.go:439
[DB:SERVIC] DEBU Received out-of-band proxy transport request for @local-node [c5e69ab0-cc11-4801-a9ad-37a91aae4064.teleport.example.net]. leaseID:1 target:teleport.example.net:3024 reversetunnel/transport.go:151
[DB:SERVIC] DEBU Handing off connection to a local "db" service. leaseID:1 target:teleport.example.net:3024 reversetunnel/transport.go:226
[DB:SERVIC] DEBU Accepted connection. addr:172.31.92.97:3024 db/server.go:343
[AUTH]      DEBU ClientCertPool -> cert(teleport.example.net issued by teleport.example.net:213176148351996403145706396176361980174) auth/middleware.go:593
[AUTH]      DEBU ClientCertPool -> cert(teleport.example.net issued by teleport.example.net:236906433016256183234006940408889049287) auth/middleware.go:593
[DB:SERVIC] DEBU Client identity: tlsca.Identity{Username:"alexmv", Impersonator:"", Groups:[]string{"admin", "editor", "test-db-access", "ssh-access-to-all"}, Usage:[]string{"usage:db"}, Principals:[]string{"root", "ubuntu"}, KubernetesGroups:[]string(nil), KubernetesUsers:[]string(nil), Expires:time.Time{wall:0x0, ext:63758506689, loc:(*time.Location)(nil)}, RouteToCluster:"teleport.example.net", KubernetesCluster:"", Traits:wrappers.Traits{"logins":[]string{"root", "ubuntu"}}, RouteToApp:tlsca.RouteToApp{SessionID:"", PublicAddr:"", ClusterName:"", Name:""}, TeleportCluster:"teleport.example.net", RouteToDatabase:tlsca.RouteToDatabase{ServiceName:"postgres-elegantly", Protocol:"postgres", Username:"", Database:""}, DatabaseNames:[]string{"example"}, DatabaseUsers:[]string{"example"}, MFAVerified:"", ClientIP:""}. db/server.go:461
[DB:SERVIC] DEBU Will connect to database "postgres-elegantly" at postgres-elegantly.example.net:5432. db/server.go:473
[DB:SERVIC] DEBU Using async streamer for session 65deb250-475b-48d7-86f5-d20b5b341e98. db/streamer.go:78
[DB:SERVIC] DEBU Received startup message: &pgproto3.StartupMessage{ProtocolVersion:0x30000, Parameters:map[string]string{"application_name":"psql", "client_encoding":"UTF8", "database":"example", "user":"example"}}. db:postgres-elegantly id:65deb250-475b-48d7-86f5-d20b5b341e98 postgres/engine.go:152
[RBAC]      DEBU Access to database "postgres-elegantly" denied, no allow rule matched. services/role.go:1932
[DB:SERVIC] ERRO Failed to handle connection. addr:172.31.92.97:3024 error:[
ERROR REPORT:
Original Error: *trace.AccessDeniedError access to database denied
Stack Trace:
	/go/src/github.com/gravitational/teleport/lib/services/role.go:1934 github.com/gravitational/teleport/lib/services.RoleSet.CheckAccessToDatabase
	/go/src/github.com/gravitational/teleport/lib/srv/db/postgres/engine.go:182 github.com/gravitational/teleport/lib/srv/db/postgres.(*Engine).checkAccess
	/go/src/github.com/gravitational/teleport/lib/srv/db/postgres/engine.go:97 github.com/gravitational/teleport/lib/srv/db/postgres.(*Engine).HandleConnection
	/go/src/github.com/gravitational/teleport/lib/srv/db/server.go:401 github.com/gravitational/teleport/lib/srv/db.(*Server).handleConnection
	/go/src/github.com/gravitational/teleport/lib/srv/db/server.go:367 github.com/gravitational/teleport/lib/srv/db.(*Server).HandleConnection
	/go/src/github.com/gravitational/teleport/lib/reversetunnel/transport.go:234 github.com/gravitational/teleport/lib/reversetunnel.(*transport).start
	/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit
User Message: access to database denied] db/server.go:369
[AUDIT]     INFO db.session.start cluster_name:teleport.example.net code:TDB00W db_name:example db_protocol:postgres db_service:postgres-elegantly db_uri:postgres-elegantly.example.net:5432 db_user:example ei:0 error:[access to database denied] event:db.session.start message:access to database denied namespace:default server_id:c5e69ab0-cc11-4801-a9ad-37a91aae4064 sid:65deb250-475b-48d7-86f5-d20b5b341e98 success:false time:2021-06-05T04:18:09.433Z uid:34f1ea7c-4a09-4ef3-bd07-a52b82a1f6c6 user:alexmv events/emitter.go:323
```

### PR DESCRIPTION
**Pull request by GavinFrazar**
Fixes https://github.com/gravitational/teleport/issues/7198

This PR updates the database retrieved for session RBAC checking (at connection time) to includes cloud and dynamic labels.

It also fixes some suspicious recursive read locking in the app server that could have caused a deadlock.

### CODE REVIEW

File: lib/srv/db/ca.go [hunk 1]
```diff
@@ -59,13 +59,45 @@ func (s *Server) startCARenewer(ctx context.Context) {
...
+	s.mu.RLock()
+	if !s.shouldInitCACertLocked(database) {
+		s.mu.RUnlock()
+		return nil
+	}
+	// make a copy so we can safely unlock before doing expensive CA cert
+	// version checking or downloading.
+	copy := database.Copy()
+	s.mu.RUnlock()
+	// It's not set so download it or see if it's already downloaded.
```
Comment (codingllama) [inline lib/srv/db/ca.go:71]: This comment seems misplaced, could we either remove or fix it?

File: lib/srv/db/access_test.go [hunk 2]
```diff
@@ -99,83 +99,130 @@ func TestMain(m *testing.M) {
...
-			allowDbNames: []string{types.Wildcard},
-			allowDbUsers: []string{types.Wildcard},
-			dbName:       "postgres",
-			dbUser:       "postgres",
+			desc:          "has access to all database names and users",
+			user:          "alice",
+			role:          "admin",
+			allowDbNames:  []string{types.Wildcard},
+			allowDbUsers:  []string{types.Wildcard},
+			allowDbLabels: wildCardDBLabels,
```
Comment (codingllama) [inline lib/srv/db/access_test.go:128]: Is it necessary to retrofit the labels in all test cases? Shouldn't we keep the existing scenarios unchanged?

File: lib/srv/db/access_test.go [hunk 3]
```diff
@@ -99,83 +99,130 @@ func TestMain(m *testing.M) {
...
-			allowDbUsers: []string{"alice"},
-			dbName:       "postgres",
-			dbUser:       "postgres",
-			err:          "access to db denied",
+			desc:          "access allowed to specific user/database by static label",
+			user:          "alice",
+			role:          "admin",
+			allowDbNames:  []string{"metrics"},
+			allowDbUsers:  []string{"alice"},
+			allowDbLabels: staticDBLabels,
```
Comment (codingllama) [inline lib/srv/db/access_test.go:181]: I'm finding it difficult to reason about this table. What is it about each entry that makes it work or not? Could we add a short comment to the new entries at least?

**Review by codingllama — APPROVED** [review]
I'm not super familiar with these parts, so I'm taking some of the changes at face value. Seems reasonable to me.

File: lib/srv/db/ca.go (hunk 1 above)
Comment (GavinFrazar) [reply lib/srv/db/ca.go:71]: this comment was there before, and as far as I can tell doesn't correspond to what actually happens regarding ".. do not update the certificates...", and the comment inside `shouldInitCACertLocked` better explains the part about the CA being set or not.

I'll just remove this comment.

Comment (GavinFrazar) [reply lib/srv/db/ca.go]: removed comment in [abab275](https://github.com/gravitational/teleport/pull/29215/commits/abab2753ed02582bc7bf9874c32c0b51bfb8de1d)

File: lib/srv/db/access_test.go (hunk 3 above)
Comment (GavinFrazar) [reply lib/srv/db/access_test.go]: added clarifying comments to the new test cases: [01c26d1](https://github.com/gravitational/teleport/pull/29215/commits/01c26d1eea4c5bd04074b89a0ba07ff8ce858d7e)

File: lib/srv/db/ca.go (hunk 1 above)
Comment (GavinFrazar) [reply lib/srv/db/ca.go]: also, the motivation for making the `shouldInitCACertLocked` helper func was to make it easier to visually see that we handle locking/unlocking correctly.

**Review by codingllama — APPROVED** [review]
Thanks, Gavin.

Well-spotted on the RWLocks, btw.

**Review by camscale — APPROVED** [review]

File: lib/srv/db/server.go [hunk 4]
```diff
@@ -598,6 +598,42 @@ func (s *Server) getProxiedDatabases() (databases types.Databases) {
...
+// a database object.
+// The caller must invoke an RLock on `s.mu` before calling this function.
+func (s *Server) copyDatabaseWithUpdatedLabelsLocked(database types.Database) *types.DatabaseV3 {
+	// create a copy of the database to modify.
+	copy := database.Copy()
+
+	// Update dynamic labels if the database has them.
+	labels, ok := s.dynamicLabels[copy.GetName()]
+	if ok && labels != nil {
+		copy.SetDynamicLabels(labels.Get())
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/srv/db/server.go:627]: This PR solves issue #28390. Let's close my PR in favor of this (since it is already solving the data race problems).

**Pull request merged**